refactor JobLog repository to doctrine

kardi31 opened the pull request:
1. Changed JobLog repository to use doctrine. 
2. Added some PHP 7.1 types
3. Improved output message on failure
4. Added Pagerfanta usage for pagination purposes

Functionally tested on dev. It requires only migrations from clients to upgrade. 

https://github.com/usemarkup/JobQueueBundle/issues/38

calumbrodie commented:
looks good, nice work. Have you tested the UI on the various screens that are in the job management area? Do they all still work?

calumbrodie commented:
e.g does the 'RabbitMq is not responsive, the service has crashed or has not been started' error still appear? I thought @gsdevme had found the cause of that...

kardi31 commented:
Good shout, one of the screens is broken. I also found search bar being 💩 so I will refactor it a bit and come back once it's done.

gsdevme commented on Resources/config/doctrine/JobLog.orm.yml:
```diff
@@ -0,0 +1,37 @@
+    exitCode:
+      type: string
+      length: 60
+      nullable: true
```
ExitCode likely shouldnt be nullable, very command has an exit code.

kardi31 replied:
it's an exit code of exception that we store here

gsdevme commented on Resources/config/doctrine/JobLog.orm.yml:
```diff
@@ -0,0 +1,37 @@
+      type: string
+      unique: true
+      length: 100
+  fields:
```
Any indexes on this table?

gsdevme commented on Resources/config/doctrine/JobLog.orm.yml:
```diff
@@ -0,0 +1,37 @@
+      type: string
+      length: 60
+    output:
+      type: string
```
Also output likely better being nullable.

gsdevme commented on Resources/config/doctrine/JobLog.orm.yml:
```diff
@@ -0,0 +1,37 @@
+    peakMemoryUse:
+      type: integer
+      length: 255
+      nullable: true
```
How much come peak memory is nullable, do we not always have a reading for the memory?

kardi31 replied:
We log jobs which are in pending state as well. They don't have any memory logged before they are executed

gsdevme commented:
> e.g does the 'RabbitMq is not responsive, the service has crashed or has not been started' error still appear?

@calumbrodie @kardi31 thats fixed already outside of this PR

kardi31 commented on Resources/config/doctrine/JobLog.orm.yml:
```diff
@@ -0,0 +1,37 @@
+  table: job_log
+  id:
+    uuid:
+      type: string
```
GUID?

kardi31 replied:
What do you mean by GUID? Also note it's a UUID of console command, not Doctrine's UUID. They are not the same

gsdevme replied:
As in there is a column type of GUID which is more fitting for storing a UUID then a string of 100 length. Remove length and change the type to GUID

kardi31 commented:
I addressed pretty much all of the feedback. 

I also run another bunch of functional tests for different Job tabs. I added ability to search by `command` name instead of `configuration_command_id`. Also added PHX PR to use Extended Pagerfanta to fix links in pagination when search data is submitted. Previously when going to page 2 we used to strip GET search data from search URL.

kardi31 commented:
![Screenshot 2019-05-31 at 10 02 15](https://user-images.githubusercontent.com/9740794/58694567-3090ce80-838b-11e9-80c2-6b0ee32777b5.png)

gsdevme commented on Entity/JobLog.php:
```diff
@@ -107,7 +90,7 @@ public function __construct(
         // the 'added' makes it unique
         if (!$uuid) {
-            $uuid = $this->generateUuid($command, $topic, $added);
+            $uuid = $this->generateUuid($command, $topic, $this->added->format('U'));
```
Can we change this to an actual UUID? just noticed its somekind of hash?

kardi31 replied:
That was my original idea, however we only generate UUID if it's not already defined.
We generate it here https://github.com/usemarkup/JobQueueBundle/blob/develop/EventSubscriber/LogConsoleCommandEventSubscriber.php#L66

I have no idea what's the format of it but I guess it's different from doctrine's UUID.

gsdevme replied:
It seems really complicated here, can't we use `Uuid::uuid4()->toString()` to generate it? 

>  UUID if it's not already defined.

Bit confused here, can we not make it a required? that always generate it outside 24/7 to reduce the need for such complex logic.

kardi31 replied:
Have you seen the link I pasted?

gsdevme replied:
> we only generate UUID if it's not already defined

Can we remove this part, and force it to always be defined?

gsdevme replied:
Generally having an optional argument that isnt optional is confusing, thats the goal of the comment here. Remove the internal logic (that is nasty) and always ensure the UUID comes from the outside... and do `Uuid::uuid4()->toString()` if required (outside)

kardi31 replied:
So when the CLI command is executed, how do you find respective log for it to mark it as complete / failed? 
Right now we do it by searching for log by command's uuid.

gsdevme replied:
Yeah we'd search by UUID that wouldn't change.

gsdevme replied:
I will PR to this branch a solution, might be easier Im perhaps not explaining very well.

gsdevme replied:
Ive completed the work here https://github.com/usemarkup/JobQueueBundle/pull/83

gsdevme commented on Entity/JobLog.php:
```diff
@@ -157,179 +135,112 @@ public function toCompressedArray()
+        $diff = $this->getCompleted()->diff($this->getStarted());
+
+        // convert diff to duration seconds
+        $duration = ($diff->days * 3600 * 24) + ($diff->h * 3600) + ($diff->i * 60) + $diff->s;
```
How come we had to change this? it kind of looks like it would of worked beforehand

kardi31 replied:
Because I changed `getCompleted` and `getStarted` to be a datetime instead of some random integers.

gsdevme replied:
When we say random integers we mean epochs right? but sure ok. I still not fully understand this snippet of code. I think it could be just be (will include in my PR)

```php
$completed = new DateTime();
$started = new DateTime('1 hour ago');

$diff = $completed->getTimestamp() - $started->getTimestamp();

var_dump($diff);
```

gsdevme commented:
There is a constant called REDIS in the repository that we need to remove also, bit confusing otherwise.

ChrisDBrown commented on EventSubscriber/CompleteConsoleCommandEventSubscriber.php:
```diff
@@ -70,19 +69,17 @@ public function onConsoleTerminate(ConsoleTerminateEvent $event)
+        // lookup job log and save status
+        $log = $this->jobLogRepository->findJobLog($uuid);
+        if (!$log) {
+            // if no log, ignore it
```
Missing return statement

kardi31 replied:
Thanks 👍 Updated

gsdevme replied:
Seems like we might want to introduce phpstan here, seems to be a bunch of things like this gone unnoticed.

kardi31 commented:
I uploaded this PR into GANT US redstaging. It looks like everything is working fine
https://gant-us.redstaging.co.uk/admin/jobs/logs/

gsdevme commented:
I would advise that. It’s public so nice to be nice to others also. 

Sent from my iPhone


> On 6 Jun 2019, at 17:32, ChrisDBrown <notifications@github.com> wrote:
> 
> @ChrisDBrown approved this pull request.
> 
> Assume we're doing this as a new major version?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

-- 
Markup
3 Rothesay House
134 Douglas Street
Glasgow
G2 4HF

+44 (0)141 416 
1221
https://www.usemarkup.com <https://www.usemarkup.com>
